storage/aptosdb/src/state_merkle_db.rs

```diff
@@ -286,26 +291,169 @@ impl StateMerkleDb {
+            .try_for_each(|row| {
+                ensure!(row.node_key.get_shard_id() == shard_id);
+                if previous_epoch_ending_version.is_some()
+                    && row.node_key.version() <= previous_epoch_ending_version.unwrap()
+                {
+                    batch.put::<StaleNodeIndexCrossEpochSchema>(row, &())
+                } else {
+                    // These are processed by the state merkle pruner.
+                    batch.put::<StaleNodeIndexSchema>(row, &())
+                }
+            })?;
+
+        Ok(batch)
+    }
+
+    // A non-sharded helper function accepting KV updates from all shards.
+    #[cfg(test)]
     pub fn merklize_value_set(
         &self,
         value_set: Vec<(HashValue, Option<&(HashValue, StateKey)>)>,
+        version: Version,
+        base_version: Option<Version>,
+        previous_epoch_ending_version: Option<Version>,
+    ) -> Result<(SchemaBatch, Vec<SchemaBatch>, HashValue)> {
+        let mut sharded_value_set: Vec<Vec<(HashValue, Option<&(HashValue, StateKey)>)>> =
+            Vec::new();
+        sharded_value_set.resize(NUM_STATE_SHARDS, Default::default());
+        value_set.into_iter().for_each(|(k, v)| {
+            sharded_value_set[k.nibble(0) as usize].push((k, v));
+        });
+
+        let (shard_root_nodes, sharded_batches) = (0..16)
+            .map(|shard_id| {
+                self.merklize_value_set_for_shard(
+                    shard_id,
+                    sharded_value_set[shard_id as usize].clone(),
+                    /*node_hashes=*/ None,
+                    version,
+                    base_version,
+                    base_version,
+                    previous_epoch_ending_version,
+                )
+                .unwrap()
+            })
+            .collect::<Vec<_>>()
+            .into_iter()
+            .unzip();
+
+        let (root_hash, top_levels_batch) = self.calculate_top_levels(
+            shard_root_nodes,
+            version,
+            base_version,
+            previous_epoch_ending_version,
+        )?;
+
+        Ok((top_levels_batch, sharded_batches, root_hash))
+    }
+
+    /// Calculates db updates for nodes in shard `shard_id`.
+    ///
+    /// Assumes 16 shards in total for now.
+    pub fn merklize_value_set_for_shard(
+        &self,
+        shard_id: u8,
+        value_set: Vec<(HashValue, Option<&(HashValue, StateKey)>)>,
         node_hashes: Option<&HashMap<NibblePath, HashValue>>,
         version: Version,
         base_version: Option<Version>,
+        shard_persisted_version: Option<Version>,
         previous_epoch_ending_version: Option<Version>,
-    ) -> Result<(SchemaBatch, ShardedStateMerkleSchemaBatch, HashValue)> {
-        let (new_root_hash, tree_update_batch) = {
+    ) -> Result<(Node, SchemaBatch)> {
+        if let Some(shard_persisted_version) = shard_persisted_version {
+            assert!(shard_persisted_version <= base_version.unwrap());
```

Comment:
should we use base_version.expect("xxxxx") ?

Reply:
Done